### PR TITLE
Fix: Centralise error handling

### PR DESCRIPTION
**Pull request by gregtyler**
Move error handling into new errorHandler wrapper function, so individual route handlers just return errors. Status, Auth and Redirect errors are handled specially, anything else is logged and throws a 500.

### CODE REVIEW

File: internal/server/server.go [hunk 1]
```diff
@@ -32,3 +41,70 @@ func New(logger *log.Logger, client Client, templates map[string]*template.Templ
...
+	return int(e)
+}
+
+type Handler func(w http.ResponseWriter, r *http.Request) error
+
+func errorHandler(name string, logger *log.Logger, tmpl403 Template, prefix, siriusURL string, next Handler) http.Handler {
+	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		if err := next(w, r); err != nil {
+			if status, ok := err.(StatusError); ok {
+				if status.Code() == http.StatusForbidden {
```
Comment (hawx) [inline internal/server/server.go:73]: Or do something generic like https://github.com/ministryofjustice/opg-use-an-lpa/blob/master/service-front/app/src/Common/templates/error/error.html.twig ?

Comment (hawx) [reply internal/server/server.go:73]: if we want specific text for some statuses we could handle that in the template instead

File: internal/server/server.go [hunk 2]
```diff
@@ -32,3 +41,70 @@ func New(logger *log.Logger, client Client, templates map[string]*template.Templ
...
+
+					logger.Printf("%s: %v\n", name, err)
+				}
+
+				http.Error(w, "", status.Code())
+				return
+			}
+
+			if err == sirius.ErrUnauthorized {
+				http.Redirect(w, r, siriusURL+"/auth", http.StatusFound)
```
Comment (hawx) [inline internal/server/server.go:97]: Sorry for the scope creep, but we will need to use a different `siriusURL` for this as it will end up being some random internal AWS URL that gets injected.

**Review by hawx — APPROVED** [review]
so much cleaner

File: docker-compose.yml [hunk 3]
```diff
@@ -6,3 +6,4 @@ services:
     ports: ["8888:8080"]
     environment:
       SIRIUS_URL: http://docker.for.mac.localhost:8080
+      SIRIUS_PUBLIC_URL: http://docker.for.mac.localhost:8080
```
Comment (hawx) [inline docker-compose.yml:9]: this should probably be `http://localhost:8080`

File: internal/server/change_password_test.go [hunk 4]
```diff
@@ -66,11 +66,13 @@ func TestPostChangePassword(t *testing.T) {
 	r.Header.Add("Content-Type", "application/x-www-form-urlencoded")
 	r.AddCookie(&http.Cookie{Name: "test", Value: "val"})
 
-	changePassword(nil, client, template, "/prefix", "http://sirius").ServeHTTP(w, r)
+	handler := changePassword(nil, client, template, "http://sirius")
+	err := handler(w, r)
+
+	redirect, ok := err.(RedirectError)
+	assert.True(ok)
+	assert.Equal("/my-details", redirect.To())
```
Comment (hawx) [inline internal/server/change_password_test.go:74]: you could probably just check `assert.Equal(RedirectError("/my-details"), err)`

File: internal/server/edit_my_details_test.go [hunk 5]
```diff
@@ -129,29 +129,30 @@ func TestGetEditMyDetailsNotPermitted(t *testing.T) {
...
-	editMyDetails(nil, client, template, "/prefix", "http://sirius").ServeHTTP(w, r)
+	handler := editMyDetails(nil, client, template, "http://sirius")
+	err := handler(w, r)
 
-	resp := w.Result()
-	assert.Equal(http.StatusFound, resp.StatusCode)
-	assert.Equal("/my-details", resp.Header.Get("Location"))
+	status, ok := err.(StatusError)
+	assert.True(ok)
+	assert.Equal(http.StatusForbidden, status.Code())
```
Comment (hawx) [inline internal/server/edit_my_details_test.go:137]: similar here could check equal with `StatusError(http.StatusForbidden)`

File: internal/server/server.go [hunk 6]
```diff
@@ -32,3 +41,76 @@ func New(logger *log.Logger, client Client, templates map[string]*template.Templ
...
+func (e StatusError) Code() int {
+	return int(e)
+}
+
+type Handler func(w http.ResponseWriter, r *http.Request) error
+
+func errorHandler(name string, logger *log.Logger, tmplError Template, prefix, siriusURL string, next Handler) http.Handler {
+	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		if err := next(w, r); err != nil {
+			type v struct {
```
Comment (hawx) [inline internal/server/server.go:72]: `v`! This may be too short even for Go. Also should be pulled up to the top-level instead of hiding in here

File: internal/server/server.go [hunk 7]
```diff
@@ -32,3 +41,76 @@ func New(logger *log.Logger, client Client, templates map[string]*template.Templ
...
+
+			if redirect, ok := err.(RedirectError); ok {
+				http.Redirect(w, r, prefix+redirect.To(), http.StatusFound)
+				return
+			}
+
+			code := http.StatusInternalServerError
+
+			if status, ok := err.(StatusError); ok {
+				if status.Code() == http.StatusForbidden || status.Code() == http.StatusNotFound {
```
Comment (hawx) [inline internal/server/server.go:92]: Could we just print the generic template for all other status codes?

Comment (gregtyler) [reply internal/server/server.go:93]: My logic was that something like a 400 should already be handled, but I guess that tallies: if we get to this stage and the only information we have is that a 400 occurred, then that's pretty much a server error.

**Review by hawx — APPROVED** [review]
